src/main/python/deep_qa/layers/attention/gated_attention.py

```diff
@@ -0,0 +1,99 @@
+          Represents the question as encoded by the biGRU.
+        - ``normalized_qd_attention``, the soft attention over the document and question.
+          Matrix of shape ``(batch, document length, question length)``.
+
+    Output:
+        - ``X``, a tensor of shape ``(batch, document length, biGRU hidden length)`` if the
+          gating function is ``*`` or ``+``, or ``(batch, document length, biGRU hidden length * 2)``
+          if the gating function is ``||``  This serves as a representation of each token in
+          the document.
+
+    Parameters
+    ----------
+    gating_function : string, default="*"
+        The gating function to use for modeling the interactions between the document and
+        query token. Supported gating functions are ``"*"`` for elementwise multiplication,
+        ``"+"`` for elementwise addition, and ``"||"`` for concatenation.
+
+    Notes
+    -----
+    To find out how we calculated equation 1, see the GatedAttentionReader model (roughly,
+    a ``masked_batch_dot`` and a ``masked_softmax``)
+    """
+    def __init__(self, **kwargs):
+        self.supports_masking = True
+        # We need to wait until below to actually handle this, because self.name gets set in
+        # super.__init__.
+        # allowed gating functions are "*" (multiply), "+" (sum), and "||" (concatenate)
+        self.gating_function = kwargs.pop('gating_function', "*")
+        if self.gating_function not in GATING_FUNCTIONS:
+            raise ConfigurationError("Invalid gating function "
+                                     "{}, expected one of {}".format(self.gating_function,
+                                                                     GATING_FUNCTIONS))
+
+        super(GatedAttention, self).__init__(**kwargs)
+
+    def compute_mask(self, inputs, mask=None):
+        # pylint: disable=unused-argument
+        # We do not need a mask beyond this layer.
+        return None
+
+    def get_output_shape_for(self, input_shapes):
+        return (input_shapes[0][0], input_shapes[0][1], input_shapes[0][2])
+
+    def call(self, inputs, mask=None):
+        # document matrix is of shape (batch, document length, biGRU hidden length)
+        # question matrix is of shape (batch, question length, biGRU hidden length)
+        # normalized_qd_attention is of shape (batch, document length, question length)
+        document_matrix, question_matrix, normalized_qd_attention = inputs
+
+        # question update is of shape (batch, document length, bigru hidden)
+        question_update = K.batch_dot(normalized_qd_attention, question_matrix, axes=[2, 1])
+
+        # use the gating function to calculate the new document representation
+        # which is of shape (batch, document length, biGRU hidden length)
+        if self.gating_function == "*":
+            return question_update * document_matrix
+        if self.gating_function == "+":
```

Comment:
Here, I think we need to utilise the mask from the document matrix. In the multiplication case, it's fine as the question update already has zeros where they should be, but for the concatenation and addition cases, if we don't mask here and don't pass a mask forward from this layer, I don't think this is going to be correctly masked upstream.